GPII-4273iii LFM gets repository revision and computes URL to solutions registry therein

klown opened the pull request:
Implementing part (iii) of GPII-4273 where the LFM  calls the `/revision` endpoint of the CBFM and uses the result to compose the full URL to the solutions registry in universal's github repository.

@cindyli @amb26 I'm doing this pull to see how the changes fare against CI, and I may change things further.  Certainly, something needs to be done in terms of specifying the actual file name `win32.json5`, currently hard-coded in the `gpii.flowManager.config.local.base.json5` configuration file.  It should come from somewhere in the `windows` implementation. That said, feel free to look it over and make comments if you like.

**NOTE**: This includes code changes from PR #837 (part ii of GPII-4273) and PR #836 (part i).

JIRA: https://issues.gpii.net/browse/GPII-4273

klown commented:
@cindyli @amb26 Ready for review, although I'm still looking into [how to parameterize the name of the solutions registry file](https://github.com/klown/universal/blob/GPII-4273iii/gpii/node_modules/flowManager/configs/gpii.flowManager.config.local.base.json5#L22), or if that is even necessary.

amb26 commented:
Thanks, @klown, this work is very nearly ready. The solution to your parameterisation problem - the implementation should not hard-code the required filename since this will be platform-dependent. Instead, this fetch should be implemented as an identical workalike to the already existing solutionsRegistryDataSource at https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/FlowManager.js#L45 and https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js#L73 - the client supplies the output of the platformReporter as the directModel if necessary, and the impl sorts out which file to dispense.
It would be worth trying to trace out how that.options.path in the solutionsRegistryDataSource gets set correctly since right now all I can see is https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/configs/gpii.flowManager.config.base.json5#L13 which sets it just to the base directory. How does it figure out from a site like https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/MatchMaking.js#L96 what is the required filename? @cindyli may be able to help with detective work

klown commented:
Thanks @amb26.  Re: the detective work:  I had started, and noticed that the `SolutionsRegistryDataSource` currently pulls in _all_ of the platform solutions in the solutions directory ,but something else chooses the actual file.  That's what I meant by whether it was even necessary for this to provide the file name when something else is doing that.

I'm also reminded me of something I did in 2017, making [the `PlatformReporter` context aware](https://github.com/klown/universal/blob/GPII-1939/gpii/node_modules/deviceReporter/src/DeviceReporter.js#L165), and having it provide the OS info.  I'm going to see if that has any use here ([windows specific implementation](https://github.com/klown/windows/blob/GPII-1939_GPII-2404/gpii/node_modules/platformReporter/PlatformReporter.js#L25)).  It might just be a the solution to the problem invovles the current platform/device reporter components.  We'll see.

klown commented:
@amb26 @cindyli Looks like it might be as simple as: `LFM.deviceReporter.platformReporter.reportPlatform()`

On macOS, that gives: `{ id: 'darwin', version: '18.7.0' }`, where the `id` is the actual filename, without the `.json5` extension.  I'll confirm that the `id` is `win32` when run on Windows.

amb26 commented:
Thanks @klown - but how does it actually pull in all the platform solutions? All I found was a single require statement...

klown commented:
@amb26, Someone very clever put an `index.js` in the solutions registry folder, which contains a `require()` statement for each solutions file in the folder.  The [algorithm for `require()`](https://nodejs.org/docs/latest-v10.x/api/modules.html#modules_all_together) has a rule that if there is an `index.js` in the folder required, load it.

I wonder if there is some way to use node's package manager to get the solutions out of the repository -- treat the solutions registry as a module.

amb26 commented:
@klown - well spotted! I feel sure that such a piece of deviousness could only possibly have been the evil work of KASPARNET @kaspermarkus

cindyli commented on gpii/node_modules/flowManager/src/FlowManager.js:
```diff
@@ -235,13 +245,23 @@ fluid.defaults("gpii.flowManager.local", {
             args: ["{webSocketsSettingsHandlerComponent}"]
         },
-
+        "onCreate.loadSolutionsFromRepository": {
```
It might be better to have `loadSolutionsFromRepository`(reading the SR from Github) as part of `SolutionsRegistryDataSource` so it's transparent to outsiders including flowManager.

What `SolutionsRegistryDataSource` does it:
1. fetch the SR for the current platform from Github;
2. If 1 fails, use the SR read from the universal /testData/solutions directory;
3. If 2 fails again, call `fluid.fail` and exit.

All above happens at the "onCreate" event of `SolutionsRegistryDataSource`. When the process completes and SR is ready, fires an event `SolutionRegistryReady` that gets escalated to `flowManager` and becomes one of the events triggering the aggregate event `flowManagerReady`.

klown replied:
This works for the `SRDS` used by the `LFM`.  But, the `CBFM` also has an `SRDS` sub-component.  In this case, the platform that the `CBFM` is running on is irrelevant.  The relevant platform is the one that the Morphic client is running on, and it communicates that to the `CBFM` via the `DeviceReporter`.  From this perspective, the `CBFM/SRDS` has to have access to all of the platforms' solutions registries since it can't know the client's platform.  Note that that is how the original `SRDS` worked, and I've kept that functionality for the `CBFM` case.

Even so, you mentioned that you found no use of the `SolutionsRegistryDataSource.getImpl()` function where the platform was passed in as one of its arguments.  I'm mystified, and am looking closer -- how does the client inform the CBFM what platform it (the client) is running on?

cindyli commented on gpii/node_modules/flowManager/src/FlowManager.js:
```diff
@@ -193,6 +195,12 @@ fluid.defaults("gpii.flowManager.local", {
         userErrors: {
             type: "gpii.userErrors"
+        },
+        gpiiRevisionRequester: {
```
`revisionRequester` should be a subcomponent of `repositorySolutionsLoader` as that's where it's needed.

It would be easier to search if "gpiiRevisionRequester" could be renamed to be the same as the component name "revisionRequester".

cindyli commented on gpii/node_modules/flowManager/src/FlowManager.js:
```diff
@@ -193,6 +195,12 @@ fluid.defaults("gpii.flowManager.local", {
+        gpiiRevisionRequester: {
+            type: "gpii.flowmanager.revisionRequester"
+        },
+        repositorySolutionsLoader: {
```
As said above, "repositorySolutionsLoader" could be a subcomponent of "solutionsRegistryDataSource" as it's only needed there.

cindyli commented on gpii/node_modules/flowManager/src/GpiiRevisionRequester.js:
```diff
@@ -0,0 +1,71 @@
+
+fluid.registerNamespace("gpii.flowmanager.revisionRequester");
+
+fluid.defaults("gpii.flowmanager.revisionRequester", {
```
This component can be replaced by the subcomponent `gpiiRevisionDataSource` and provide a get() API.

cindyli commented on gpii/node_modules/flowManager/src/RepositorySolutionsLoader.js:
```diff
@@ -0,0 +1,115 @@
+    // file.
+    urlPrefix: "",
+    urlSuffix: "",
+    solutionsRegistryFolderPath: "",
```
This folder path doesn't see to be use anywhere. According to your comment, this path is intended to be used to save the contents of the downloaded solutions registry to the local system. I doubt if we should ever save the downloaded SR to the local disk, especially it's pointing to `"%gpii-universal/testData/solutions"` in [the config file](https://github.com/GPII/universal/pull/840/files#diff-ce29f5ee38b24db477e76b84aeadfe6dR21-R23), which means to overwrite the SR in the universal code base.

cindyli replied:
`solutionsRegistryFolderPath` is not used in the functional code. Shall it be removed? If it can be removed, please also clean up its appearances in tests and config files. 

Please also revisit the code comment above about "save the contents of the downloaded solutions registry file". Thanks.

klown replied:
Okay removed, but it was tricky finding all of them.

cindyli commented on gpii/node_modules/flowManager/src/FlowManager.js:
```diff
@@ -358,6 +381,56 @@ gpii.flowManager.local.savePreferences = function (that, gpiiKey, preferences) {
+ *      solutions registry file from the source code repository.  It is stored
+ *      in the respositorySolutionsLoader's solutionsRegistry member.
+ */
+gpii.flowManager.local.loadSolutionsFromRepository = function (that, loadedFromRepositoryEvent) {
```
As said earlier, this logic probably makes more sense to be inside `solutionRegistryDataSource` component.

Also, by looking at nested promises in here, please consider to use [fluid.promise.fireTransformEvent()](https://docs.fluidproject.org/infusion/development/PromisesAPI.html#fluidpromisefiretransformeventevent-payload-options) that can chain up asynchronous promises and pass the output of the previous promise to the next one on the chain. 

In the `solutionRegistryDataSource` component, the sequence of actions performed by a chain could be:
1. fetch the revision from remote CBFM;
2. Use the revision to fetch SR from Github;
3. If SR from 2 is not null, pass the SR to 4; if it is null, read SR from the local disk and pass to 4;
4. If SR is not null, fire the success event. Otherwise, call `fluid.fail()`.

cindyli commented on gpii/node_modules/flowManager/src/RepositorySolutionsLoader.js:
```diff
@@ -0,0 +1,115 @@
+        solutionsRegistry: null
+    },
+    components: {
+        encoding: {
```
This `encoding` subcomponent doesn't seem to be used.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -50,15 +50,16 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
 
 // TODO: Add an invoker to reload once we are "more live".
 gpii.flowManager.solutionsRegistry.dataSource.loadSolutionsRegistry = function (that) {
```
Currently, what SolutionsRegistryDataSource.get() does is:
1. If an argument of "os" is given, returns the SR for this platform;
2. If no argument is given, returns a fulllll list of SRs including for windows, linux, darwin, android etc.

I did a search across the code base, point 2 has never been used.

As now only the SR for the platform that Morphic is running on will be fetched from the remote Github, I think it's a good time to refactor this API:
1. SolutionsRegistryDataSource will auto detect the current platform via platform detector;
2. SolutionsRegistryDataSource.get() only returns the SR for the current platform, no argument is needed.

cindyli commented on gpii/node_modules/flowManager/src/RepositorySolutionsLoader.js:
```diff
@@ -0,0 +1,115 @@
+
+fluid.defaults("gpii.flowManager.repositorySolutionsLoader", {
+    gradeNames: ["kettle.dataSource.URL"],
+    urlTemplate: "%prefix/%revision/%suffix/%fileName",
```
In the case such as in tests, if we don't want to fetch SR from the remote, we could potentially override `urlTemplate` to an empty string which will result in `url` being an empty string. The fetch function first checks if `url` is provided, if not, skip fetching from the remote. This gives us an option to skip the remote fetch.

cindyli commented:
@klown, I issued https://github.com/klown/universal/pull/8 against your branch. It should fix the failing production config test.

klown commented:
@cindyli wrote:

> Currently, what SolutionsRegistryDataSource.get() does is:
> 
>    1.  If an argument of "os" is given, returns the SR for this platform;
>    1.  If no argument is given, returns a fulllll list of SRs including for windows, linux, darwin, android etc.
> 
> I did a search across the code base, point 2 has never been used.

I found three cases where the full list of SRs is fetched.  I'm currently working through the most complex case, that involving the `MatchMaker` -- see `gpii.flowManager.getSolutions()`[https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/MatchMaking.js#L96]

It first fetches all of the solutions and then immediately filters out the one for the current platform.  That's easily fixed and replaced with a request for just the current platform's solutions.  However, in the subsequent code, the full set of solutions is forwarded on twice: once as [the payload of an event](https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/MatchMaking.js#L109), and once as [the resolved value of a promise](https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/MatchMaking.js#L106).

I've traced the event part of this and determined that the full solutions are not actually used anywhere, and could be removed from the payload.  But, I'm still working on what happens to the value of the promise.

While I'm working through this, I've implemented the two grades of the `SolutionsRegistryDataSource` as having exactly the same API.  As far as a client is concerned, it can request all of the solutions or a specific one.  The SRDS used with the LFM/gpii-app does some extra work.  If it fetched that platform's solutions from github, it _overlays_ that set of solutions:
1. if the full list of solutions is requested, they are all returned, but the one set for the current platform (from github) is substituted for whatever is on the local file system,
1. if a specific platform is requested, that one is returned, but if the request is for the current platform, the one from github is returned, not the one on the local file system.

Finally, there is an odd platform named ["`web`"](https://github.com/GPII/universal/blob/master/testData/solutions/web.json5), that doesn't seem to match any OS, and it may be a factor here.  Is it possible that even if `gpii-app` is runnning on, say, `win32`, it will still need to access the solutions for the `web` platform?  If so, the "local" SRDS will have to be able to return the `web` solutions in addition to the actual OS/platform.  I'm going to investigate that next, but if you know the importance of the `web` solutions, let me know.

@amb26^

klown commented:
> I believe web platform is a legacy keyword ...

Thanks @cindyli.  Maybe we can just delete it.

cindyli commented:
> > I believe web platform is a legacy keyword ...
> 
> Thanks @cindyli. Maybe we can just delete it.

I do find some solutions in this file appear in browser channel tests. To delete it, a thorough assessment and QA is needed.

klown commented:
I wrote:
>  I'm currently working through the most complex case, that involving the MatchMaker -- see [`gpii.flowManager.getSolutions()`](https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/MatchMaking.js#L96)

There is a call inside `gpii.flowManager.getSolutions()` to `solutionsRegistryDataSource.get({});`, which is the way to get the full list of solutions for all platforms, currently.   Here how that's used:

- the resolution of the `promiseToGo` that follows contains the full solutions list as a member, but that is not subsequently used.
- the `onSuccessEvent` fires with a payload with the full list as a member, but that is also not subsequently used.
- the `JournalIntegrationTests` [define a listener for the `onRead` event](https://github.com/GPII/universal/blob/master/tests/JournalIntegrationTests.js#L238) of the `solutionsRegistryDataSource (SRDS)` to intercept the payload and [overlay it with test data](https://github.com/GPII/universal/blob/master/tests/JournalIntegrationTests.js#L357) that modifies the solutions for the `win32` platform.

Of these three, only the last one depends on the full solutions, but it need not.  It only modifies the `win32` solutions, and then uses only those solutions in its test cases.  It can be re-written to handle the case where the solutions registry data source returns only one platform.

Related:  There is a another use of the full list of solutions in the [`CaptureTests`](https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/test/CaptureTests.js).  But, here, there is no reliance on the API; instead, the code directly accesses the `fullSolutionsRegistry` member of the `SRDS` component, modifies it, and then [sets it back](https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/test/CaptureTests.js#L56).  The tests then focus on only one platform.  I'm of two minds here:  this is simply setting up solutions data for the tests, so it could stay as is.  A "purer" approach would be to use the same overlay technique that the `JournalIntegrationTests.js` do.

klown commented:
Ready for review @cindyli @amb26

cindyli commented on gpii/node_modules/flowManager/src/CloudBasedFlowManager.js:
```diff
@@ -24,6 +24,8 @@ require("./HealthGetHandler.js");
 require("./ReadyGetHandler.js");
 require("./SettingsGetHandler.js");
 require("./SettingsPutHandler.js");
+require("./SettingsPutHandler.js");
```
This line is a duplicate of line 26 and can be removed.

cindyli commented on gpii/node_modules/flowManager/src/FlowManager.js:
```diff
@@ -318,13 +332,10 @@ gpii.flowManager.local.resetAtStart = function (lifecycleManager, resetAtStart,
     defaultSettingsDataPromise.then(function (defaultSettingsData) {
         var defaultSnapshot = defaultSettingsData.defaultSnapshot;
         if (resetAtStart && defaultSnapshot) {
-            fluid.log("Resetting on system startup to default settings: ", defaultSnapshot);
```
These removed logging lines at 321, 324, 327 are useful to report if the reset at start is performed and the result. Could you not remove them? Thanks.

klown replied:
Oops.  I had repurposed them for my own "TRACE:" debugging messages, and when I search and removed those, these matched.  Thanks for catching that.

cindyli commented on gpii/node_modules/flowManager/src/RevisionGetHandler.js:
```diff
@@ -0,0 +1,55 @@
+    var fullPath = fluid.module.resolvePath(gpiiRevisionPath);
+    var result = {sha256: ""};
+    try {
+        result = JSON.parse(fs.readFileSync(fullPath, "utf-8"));
```
It would be worthwhile to check the existence of `fullPath` before reading it and expecting the `catch` to report the error.

As `gpii-revision.json` doesn't exist in the universal root directory, right now running tests populates the log with fatal errors although tests will complete successfully:
```
15:30:13.955:  FATAL ERROR: Uncaught exception: Error retrieving full git revision: ENOENT: no such file or directory, open '/home/vagrant/sync/universal/gpii-revision.json'
undefined
...
```

klown replied:
@cindyli This is actually part of PR #837 that has only the changes for the CBFM `/revision` endpoint.  I'll fix it there and merge it into this one.

klown replied:
I started on this, but there is a bigger issue. 

This is the `/revision` GET handler, and like other GET handlers, it is either successful, or there is an error, here a `404`.  The success/failure is implemented in the handler by firing either an `onSuccess` or `onError` event.

Thus, checking if the file exists doesn't help, since a `200` (success) is not correct if the file is missing. The `onError` event is still going to fire, and will result in a "`FATAL ERROR: ...`" message in the tests.

I'm wondering if this is peculiar to the tests, and whetehr there is a way to avoid the error for cases where getting the revision is not relevant to a specific test.

cindyli replied:
It seems [`fs.access`](https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback) or `fs.accessSync` check the file existence without opening it, so they don't report fatal errors. Please give a try. Thanks.

klown replied:
Using `fs.accessSync()` doesn't help.

The _reporting_ of the `FATAL ERROR: ...` is not caused by the failure of `fs.readFileSync()` or either of the access functions.  It's because that failure is relayed to an `onError` event a few lines later, using the call to `gpii.flowManager.cloudBased.gpiiRevision.handler.reportError()`.

Both of the "missing revision file" and "malformed revision file" errors result in a `FATAL...` log message.  The latter produces messages like:
> `17:34:19.500:  FATAL ERROR: Uncaught exception: Error retrieving full git revision: Missing revision value`

Note that both error conditions are tested in the `CloudRevisionGetTests.js` in the test cases [`testInvalidRevision`](https://github.com/klown/universal/blob/GPII-4273iii/tests/platform/cloud/CloudRevisionGetTests.js#L86) and [`testMissingRevision`](https://github.com/klown/universal/blob/GPII-4273iii/tests/platform/cloud/CloudRevisionGetTests.js#L98)

The use of `onError` for requests is used in many places in the GPII.  For example, the `/ready` and `/health` endpoints output `FATAL...` log messages when they fail, and there are others.  The reason there are so many in the case of `/revision` is because the new local SRDS always requests the revision when it initializes, and gpii-universal doesn't have such a file in the repository. So, for every test that causes the creation of a local SRDS, this error will occur. Which is, admittedly, a lot.  But, as you noted, it's harmless in that is does not interfere with any given test, just like all the other instances of `FATAL ...` log messages.

I'm exploring if there is a way to configure the tests such that revision requests are omitted where that's not important to the tests.

cindyli replied:
Good point of omitting revision requests in tests. This probably makes sense for running the system in dev configs too, in which case, people or developers who are testing the system might prefer to read the in-dev solutions registry directly from the file system.

My thought is:
1. Move [configs for the SR remote fetch](https://github.com/klown/universal/blob/GPII-4273iii/gpii/node_modules/flowManager/configs/gpii.flowManager.config.local.base.json5#L9-L20) out of the `gpii.flowManager.config.local.base` into its own separate config file. This config will included into production configs in "gpii/configs/" directory. An example of such a separate config is [here](https://github.com/klown/universal/blob/GPII-4273iii/gpii/node_modules/flowManager/configs/gpii.flowManager.config.suppressHttpEndpoints.base.json5). If we do so, a documentation should be added for this config at [configs.md](https://github.com/klown/universal/blob/GPII-4273iii/documentation/Configs.md#special-configs);

2. In this separate config, the distribution target for `cloudURL`, `urlPrefix`, `urlSuffix` will be `gpii.flowManager.solutionsRegistry.dataSource.local` rather than its sub-components `revisionRequester` and `repositorySolutionsLoader`, so that [`loadSolutions.getRevision` listener](https://github.com/klown/universal/blob/GPII-4273iii/gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js#L174-L177) can verify `cloudURL !== null` before sending requests to CBFM and [`loadSolutions.loadFromRepository`] can verify `urlPrefix` before fetching from Github repo. Extra `distributeOptions` blocks will be added to `gpii.flowManager.solutionsRegistry.dataSource.local` to distribute these 3 values down to its subcomponents;

3. Add a test config for testing SR fetch from Github repo.

Let me know if you come up with better solutions. Thanks.

klown replied:
The "problem" with the `FATAL ...` messages in the log for some of the tests
is twofold:

1. Some of the tests instatiate a CBFM that supports the `/revision` endpoint.  AND they instantiate a LFM that uses the new `SRDS` that, in turn, queries that CBFM `/revision` endpoint.  The CBFM's revision request handler will fire an `onError` event in this case because there is no revision file for the CBFM to return.  The error event is surfaced as a `FATAL ...` error message.  This is the way it is supposed to work.  For example, by comparison, the production tests do not show the error because they are set up to mimic a production system and, as such, have a revision file.  If the revision file was missing or malformed in a production scenario, you'd want to see the `FATAL` error messages.

2. The testing framework defines a fluid component whose job is to ensure that `FATAL` errors are logged.  The comment about the [`errorLoggingRequest` component](https://github.com/GPII/universal/blob/master/gpii/node_modules/testing/index.js#L8) indicates that this is a requirement: 

> `Some definitions to make sure that errors during request get logged to the console. This should really be a default`

I removed that component as an experiment and none of the `FATAL` error messages are logged, but otherwise everything else behaves as before.

The dilemma is that the testing framework sees the `FATAL` error messages as required, but the ones caused by a CBFM-with-revision-endpoint + a LFM-with-new-SRDS are irrelevant for the majority of the tests.

The compromise I came up with is to defeat the revision request, but in just those cases where it is pointless.  This is done by changing a couple of the configuration file to misdirect the `/revision` request away from the CBFM.  That is, when the SRDS asks for the revision, it asks that of a server that has no `/revision` endpoint and, in pariticular, no revision request handler.  The `onError` event won't occur.  However, for the specific tests of the `/revision` endpoint (`CloudRevisionGetTests.js`, and the production tests), the configuration is correctly set and the `FATAL` messages are output.  Furthermore, any other tests that do incur a `FATAL` error continue to work as before.  Those `FATAL` errors are logged.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
 };
 
+// Solutions registry data source used by the Cloud Based Flow Manager
+fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource.cloud", {
```
We probably should name this component `gpii.flowManager.solutionsRegistry.dataSource.cloudBased` (cloud -> cloudBased) to match the the name of Cloud Based Flow Manager `gpii.flowManager.cloudBased`.

klown replied:
Okay

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
- * Handler for get requests of solutions registry. It will return either a full solution registry,
- * or if an 'os' is provided in the requestOptions, only the entries for that os will be returned
+ * Handler for get requests to the solutions registry used by the Cloud Based
+ * Flow Manager. It will return either an object containing solutions registries
```
The sentence of "It will return either ..." is incomplete missing "or ..." part. :)

klown replied:
LOL.  There is no "or" part.  The "either" should not be there.  Still, it needs a little grammar work.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
+    return promise;
+};
+
+// Solutions registry datasource used by the Local Flow Manager.  It loads both
```
It would be nice if you can expand on the second sentence with the workflow and fallback. Thanks.

klown replied:
Done.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
+        }
+        // The platformReporter subcomponent is distributed down from local
+        // flowManager
+//      platformReporter: {
```
Place the double slashes (//) at line 147-149 to be at the same spot as line 145-146.

klown replied:
Okay

cindyli replied:
Bump. Line up indentations of commented lines with lines around this chunk.

cindyli commented on gpii/node_modules/flowManager/src/RepositorySolutionsLoader.js:
```diff
@@ -0,0 +1,110 @@
+    invokers: {
+        getSolutions: {
+            funcName: "gpii.flowManager.repositorySolutionsLoader.getSolutions",
+            args: ["{that}", "{arguments}.0", "{arguments}.1"]
```
`gpii.flowManager.repositorySolutionsLoader.getSolutions()` function at line 80 accepts 4 parameters. This argument list only passes the first 3. Shall the fourth be added here? Does this also mean the test doesn't cover the fourth argument?

cindyli replied:
Moreover, due to the connection between the platform id and the file name: file-name === platform-id + ".json5". Passing platform id might be good enough.

klown replied:
As it happens, the `platformId` is actually part of the `fileName`.  For example, if the solutions registry file to download is `win32.json`, then the `platformId` is `win32`. Passing the platform id is simply a convenience.  I'll re-write `getSolutions()` to get the platform id from the `fileName`.

klown replied:
Somehow, I completely missed your excellent suggestion of just passing a `platformId`, and constructing the `fileName` inside of `getSolutions()`.  That's much more intuitive, and I've rewritten it that way.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
+ * If successful, sets the repositorySolutionsRegistry member to the fetched
+ * solutions registry.
+ * @param {Component} that - An instance of gpii.flowManager.solutionsRegistry.dataSource
+ * @param {Component} that.repositorySolutionsLoader - An instance of
```
The parameters described at line 189 and 190 are not in the argument list. These descriptions can be removed.

klown replied:
Done.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,197 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
+        // Either the solutions registry has been successfully retrieved from the
+        // repository, or it hasn't (= null).  Set that.repositorySolutionsRegistry
+        // to whatever the result.
+        that.repositorySolutionsRegistry = that.repositorySolutionsLoader.solutionsRegistry;
```
It probably makes more sense to have `that.repositorySolutionsLoader.getSolutions` return a promise whose resolved value is the expected solutions registry, rather than reaching into `that.repositorySolutionsLoader.solutionsRegistry`.

klown replied:
Okay.  But as I fixed it, I noticed that the RepositorySolutionLoader was using `resolve()` for the error conditions.  I've switched that to `reject()`, and adjusted its tests and the SolutionsRegistryDataSource appropriately.

cindyli commented on gpii/node_modules/flowManager/test/CaptureTests.js:
```diff
@@ -47,22 +47,31 @@ gpii.tests.flowManager.capture.platformReporter = function () {
 
-gpii.tests.flowManager.capture.adjustSolutions = function (solutionsRegistryDataSource) {
+gpii.tests.flowManager.capture.adjustSolutions = function (solutionsRegistryDataSource, revision, platformReporter) {
+    var platformId = fluid.invokeGlobalFunction(platformReporter).id;
```
This test has defined the use of "darwin" platform at line 46. I think it's unnecessary to have line 50-51 and fetch this known id from `gpii.tests.flowManager.capture.platformReporter`. To simplify, you could define `gpii.tests.flowManager.capture.platformId = "darwin"` at the beginning of the test file and use this variable whenever needed.

klown replied:
Okay

cindyli commented on gpii/node_modules/flowManager/test/CaptureTests.js:
```diff
@@ -72,10 +81,18 @@ fluid.defaults("gpii.tests.flowManager.capture.tests", {
                 sequence: [{
-                    funcName: "gpii.tests.flowManager.capture.adjustSolutions",
-                    args: ["{config}.server.flowManager.solutionsRegistryDataSource"]
+                    event: "{config}.server.flowManager.solutionsRegistryDataSource.events.solutionsRegistryReady",
```
line 86-95 is to verify the solution registry has been populated correctly, which is not a part of testing the capture tool. If these events are in place to make sure the capture test starts when SR is ready, having one of these events in the test sequence should be enough.

klown replied:
Yes, it's important is to wait until the SRDS creation sequence is done before running the tests. Given that, I've kept the solutionRegistryReady as the appropriate event for these tests.

cindyli commented on gpii/node_modules/flowManager/test/SolutionsRegistryDataSourceTests.js:
```diff
@@ -0,0 +1,320 @@
+            sequence: [
+                { funcName: "fluid.log", args: ["Solutions Registry for CBFM, getting named registry -- START"] },
+
+                // For some reason, even though the solutionsRegistryReady event
```
Shall these comments and commented code be removed?

klown replied:
I was hoping the reviewers would have some insights here :-)

I currently think that by the time the tests are marshalled and ready to go, the event is long  gone, but I'm not sure.  In any case, I'll remove the comments and code now.

klown replied:
Also, since I now understand the `task/resolve/reject` pattern in the test fixtures better, I've rewritten them to no longer depend on the `onRead` event.  Using the `task` pattern to check the promise returned by a `SRDS.get(...)` is what is needed here.

cindyli commented on gpii/node_modules/flowManager/test/SolutionsRegistryDataSourceTests.js:
```diff
@@ -0,0 +1,320 @@
+            listener: "{revisionRequester}.getRevision",
+            priority: "after:loadFromLocalDisk"
+        },
+        "loadSolutions.loadFromRepository": {
```
This listener appears the only one required for overriding the base component. Please consider to remove other 3 listeners so when they are modified in the base component. These 3 here won't unintentionally override ones from the base.

klown replied:
I tried that initially, and nothing worked.  Then, I tried different combinations of the listeners until I ended up replicating the entire set from the base grade.  However, I've since tried your suggestion, and it's fine.  It must have been a typo, before, I guess.  Upshot:  removed all but the critical listener.

cindyli commented on gpii/node_modules/flowManager/test/SolutionsRegistryDataSourceTests.js:
```diff
@@ -0,0 +1,320 @@
+                    listener: "jqUnit.assert",
+                    args: ["LFM solutionsRegistryMockRepoLoad ready"]
+                }, {
+                    funcName: "{solutionsRegistryMockRepoLoad}.get",
```
Is the second `get()` call at line 187 needed? The result from this `get()` call can be verified for both: `darwin` content matches what's received from the repo and `win32` content matches what's in the local file system.

klown replied:
Yes, it's needed.  The result of the `get()` at line 172 is just the `darwin` solutions.  Specifically, the promise returned by the `get()` resolves to an object containing all and only `darwin` solutions.  Nothing is returned about the `win32` content.  And, as you noted, this call to `get()` is to check that the `darwin` solutions are the ones retrieved from the source code repository, not the ones from the local file system.  I've modified the 
"check" function to make this more apparent.

The call to `get()` on line 187 is for checking that the `win32` solutions are returned, _and_ that they match the ones on the local file system.

That said, these tests are only somewhat meaningful.  They do check the logic of which solutions to return. But, by the time the `get()` requests are made, all of the solutions registries are in memory inside the SRDS.  Most test failures here point to prior failures of the solutions loading sequence.

Related:  Since the solutions are all in memory, why is this a GET request?  The result could be an immediate return of solutions rather than a fulfilled promise.  Is there a requirement that this `get()` be RESTful?  Perhaps @amb26 knows/remembers.

cindyli commented on gpii/node_modules/flowManager/test/SolutionsRegistryDataSourceTests.js:
```diff
@@ -0,0 +1,320 @@
+                        "{arguments}.0"
+                    ]
+                },
+                { funcName: "fluid.log", args: ["Solutions Registry Mock Repo Load, getting all registries -- START"] }
```
I believe you meant `END` for logging. ;)

klown replied:
Yes.  Copy/paste is your friend ... and your worst enemy.

cindyli commented on gpii/node_modules/flowManager/test/SolutionsRegistryDataSourceTests.js:
```diff
@@ -0,0 +1,320 @@
+            listener: "{revisionRequester}.getRevision",
+            priority: "after:loadFromLocalDisk"
+        },
+        "loadSolutions.loadFromRepository": {
```
Same as a previous comment. This listener should be the only one kept for overriding the corresponding listener in the base component. Other 3 that are identical with the base component can be removed from here.

klown replied:
Right.

cindyli commented on gpii/node_modules/flowManager/test/SolutionsRegistryDataSourceTests.js:
```diff
@@ -0,0 +1,320 @@
+                    listener: "jqUnit.assert",
+                    args: ["LFM solutionsRegistryNullRepoLoad ready"]
+                }, {
+                    funcName: "{solutionsRegistryNullRepoLoad}.get",
```
Same as a previous comment too. One set of `get()` call and the verification should be enough, rather than 2 sets.

klown replied:
Same reply as above.  And, in this case, since it mocks a failure to load solutions from the source code repository, this test is checking the fallback of using the ones on the local file system.  It's arguable that the second call to `get()` is redundant since it's testing the same thing as the second `get()` in the mock repository load test case, above.

cindyli commented on gpii/node_modules/flowManager/test/shared/BrowserChannelTestDefs.js:
```diff
@@ -53,6 +53,15 @@ gpii.tests.flowManager.browserChannel.reportPlatform = function () {
+// code repository in order to use the canned registry set up for these tests.
+gpii.tests.flowManager.browserChannel.getRepositorySolutions = function (repositorySolutionsLoader) {
+    var togo = fluid.promise();
+    repositorySolutionsLoader.solutionsRegistry = null;
```
This is another indication that having `repositorySolutionsLoader.solutionsRegistry` to track fetched SR may not be a good idea. Returning the SR in the promise returned by `get()` should be enough. Having 2 reaching points would confuse other developers when both points need to be taken care of when they'd like to do a similar overriding.

klown replied:
RIght, and given I've change `RepositorySolutionsLoader` to return a promise to the`SRDS`, Ive re-worked this as appropriate.

cindyli commented:
Can you add documentation about the workflow of the fetch of the solutions registry. Maybe rename ["documentation/SolutionsRegistryFormat.md"](https://github.com/GPII/universal/blob/master/documentation/SolutionsRegistryFormat.md) to "documentation/SolutionsRegistry.md" and add a section for it. Thanks.

klown commented:
> Can you add documentation about the workflow of the fetch of the solutions registry. Maybe rename ["documentation/SolutionsRegistryFormat.md"](https://github.com/GPII/universal/blob/master/documentation/SolutionsRegistryFormat.md) to "documentation/SolutionsRegistry.md" and add a section for it. Thanks.

I decided that" SolutionsRegistryFormat.md" was really about the format of the registry, and it is fairly full already.  I decided to create a separate "SolutionsRegistryDataSource.md" in the documentation folder.  Just like there are no tests of the SRDS in master, there is no documentation about it either.  I added an overview of this component as well as the new workflow.

klown commented:
Ready for review.

Note:  with the addition of the new production configuration for the LFM (to request the GPII revision from the CBFM), I suspect that the configuration to use with [`gpii-app`](https://github.com/GPII/gpii-app/blob/master/configs/app.production.json) changes as well.  I'll be looking into that.

cindyli commented on documentation/SolutionsRegistryDataSource.md:
```diff
@@ -0,0 +1,122 @@
+depending on whether the flow manager is running in the cloud or locally on the
+client device.  This are described in the following two sections.
+
+## Cloud Based Flow Manager
```
Nice writing. Thanks, @klown.

The section title of "Cloud Based Flow Manager" might be misleading. What about "Cloud Based Solution Registry Data Source"? The section content then describes it is running as a sub-component in the Cloud Based Flow Manager.

The same comment applies to the next section title "Local Flow Manager".

klown replied:
The paragraph that precedes the section is meant to introduce the next two sections as two different scenarios and describe the way the SRDS is initialized in the context of the CBFM and the LFM. I've tried to make that more explicit, and changed the headings.  Hopefully it's clearer.

cindyli commented on documentation/Configs.md:
```diff
@@ -84,3 +84,12 @@ for the reset API and how to use it.
+
+### Handle Revision Endpoint in Production
+
+**Config file**: [`%universal/gpii/configs/gpii.config.production.local.flowManager.revision.json5`](../gpii/configs/gpii.config.production.local.flowManager.revision.json5)
```
A couple of thoughts:
1. This config should be an add-on config that will be up to config creators to add when they need to. We probably don't want to create a full untrusted config.
2. Based on 1, this config should be in `gpii/node_modules/flowManager/configs` with a name `gpii.flowManager.config.untrusted.remoteSolutionRegistry.base.json5`. It will decide more than whether to fetch revision hash from the Cloud, which is just one step of the remote SR fetching. It determines whether the solutions registry will be fetched from the Cloud.

klown replied:
Done.

cindyli commented on gpii/configs/gpii.config.production.local.flowManager.revision.json5:
```diff
@@ -0,0 +1,24 @@
+            }
+        }
+    },
+    "mergeConfigs": [
```
According to the previous comment, this `mergeConfigs` block can be removed.

cindyli commented on gpii/configs/gpii.config.production.local.flowManager.revision.json5:
```diff
@@ -0,0 +1,24 @@
+        "distributeOptions": {
+            "flowManager.defaultRevisionCloudURL": {
+                "record": "http://localhost:8084",
+                "target": "{that flowManager solutionsRegistryDataSource revisionRequester}.options.cloudURL"
```
The `flowManager` in the target should be `gpii.flowManager.local` to be specific on distributing this option to LFM. Same for line 16.

klown replied:
Right.

klown replied:
Followup:   This distribution was originally part of "gpii.flowManager.config.untrusted.base.json5" in the flowmanager configs.  I used the record/target for the [settings data source](
https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/configs/gpii.flowManager.config.untrusted.base.json5#L14) as the basis for revision requester's cloud url.  It also uses the expression `"{that flowManager ...}"` to pick out the LFM.  I have no objection to using "gpii.flowManager.local" here but is the other use okay?

cindyli replied:
`gpii.flowManager.config.untrusted.base.json5` only deals with the local flow manager so there isn't a risk to accidentally distribute to both LFM and CBFM, but it's always better to be specific with the intended target.

cindyli commented on gpii/node_modules/flowManager/src/GpiiRevisionRequester.js:
```diff
@@ -0,0 +1,84 @@
+        togo.resolve({
+            isError: true,
+            message: "No valid '/revision' endpoint, ignoring request",
+            statusCode: 404
```
No need to return `statusCode: 404` which only makes sense in the cloud http request handler components that need to return a http status code.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,209 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
+        promise.reject({
+            isError: true,
+            message: "The requested OS (" + requestOptions.os + ") was not present in the solutions registry",
+            statusCode: 404
```
Same as the comment above, don't need to return `statusCode: 404`. Same comment applies to line 254, 261.

klown replied:
Okay, but this one (L129) is old code from the original commit back in 2016:
https://github.com/GPII/universal/blob/f1c2eee4c99e06a60a06dcf78147f3bbe22d5f20/gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js#L75

As far as I can tell, the authors' intent was to simulate a GET request/response paradigm and that paradigm includes status codes.  But, as I [wrote above -- last paragraph](https://github.com/GPII/universal/pull/840#discussion_r395760908), I don't see a reason for doing this since by the time the function is called, the information is available and there is no need to wrap it in a promise.

In any case, I removed the status code.

cindyli commented on gpii/node_modules/flowManager/src/SolutionsRegistryDataSource.js:
```diff
@@ -35,45 +43,209 @@ fluid.defaults("gpii.flowManager.solutionsRegistry.dataSource", {
+ *      registry from the source code repository, or null, if there was an error.
+ */
+gpii.flowManager.solutionsRegistry.dataSource.local.loadFromRepository = function (that, revision, platformReporter) {
+    var gpiiRevision = revision.sha256;
```
At the start of the process, it might be worthwhile to check if a proper `revision` is received. If `revision` is not received due to `cloudURL === null`, return and skip the rest.

cindyli commented on gpii/node_modules/flowManager/src/GpiiRevisionRequester.js:
```diff
@@ -0,0 +1,84 @@
+        // If the url to the CBFM is null, assume this is running in a
+        // development testing enviroment where there is no valid /revision
+        // endpoint.
+        togo.resolve({
```
After thinking a bit more about it, rather than resolving an error message, it might make more sense to `togo.resolve(null)` as `cloudURL === null` is a intentional and supported case.

In the documentation, please add that setting cloudURL to `null` is the approach to stop LFM fetch SR from the Cloud but using local SR instead.

cindyli replied:
Add to the above, `gpii.flowManager.solutionsRegistry.dataSource.local.loadFromRepository` will look like:
```
if (!revision) {
    togo.resolve(null);
} else {
    fetch SR from Cloud; 
}
```

klown replied:
At the risk of derailing myself:  I am and have been uncomfortable using `resolve()` here instead of `reject()`.  After all, it is a failure and normally that means rejecting the promise.  However, the problem with `reject()` is that the sequence would come to a halt, or at least that's my understanding of what happens with `fluid.promise.fireTransformEvent()` -- each transform is executed until one of them is rejected, or until all of them are complete.  In the case of the SRDS, the `solutionsRegistryReady` event is fired by the last step, but only if the whole sequence is successful (?).

Given that, would a reject() be an option if there were a way to include a rejection handler for the whole sequence, that fires the required `solutionsRegistryReady` event?  If that were possible, a single reject() would abort the rest of the sequence, and there is no need for resolving to null (as you suggest), or to a payload with an error and/or status code (as I've written it).  The rejection would invoke this proposed handler to call "SRDS.events.solutionsRegistryReady".  Or is that just not how it works with a transform event sequence?

I'll explore it a bit.

cindyli replied:
My understanding is the same that the rejection here will cause the `fluid.promise.fireTransformEvent()` chain to reject from here. It won't proceed with the rest actions.

I feel using `resolve()` makes sense here because `url === null` is the way-out case we support for telling the system not to fetch from remote but use the local instead.

klown replied:
Okay, but FWIW, I took a close look at how `fluid.promise.fireTransformEvent()` works.  The call returns a `promise` that can be thought of as the promise for the whole sequence.  A rejection handler can be added to it that, in this case, fires the `solutionsRegistryReady` event.  Then, a rejection at any point in the load sequence, (1) ends the sequence immediately, (2) fires the "solutionsRegistryReady" event.

Having said that, I didn't use it, and went with a version of your suggestion to test the revision, and use `resolve()`.  I did, however, test the idea and, if you are interested, have a look at this gist: https://gist.github.com/klown/a4f13aa6b2207e82d403c38ffb20ecd1

cindyli commented on gpii/node_modules/flowManager/src/RepositorySolutionsLoader.js:
```diff
@@ -0,0 +1,108 @@
+        getSolutions: {
+            funcName: "gpii.flowManager.repositorySolutionsLoader.getSolutions",
+            args: ["{that}", "{arguments}.0", "{arguments}.1"]
+                             // gpii revision, file name
```
"file name" -> "platform id"

klown replied:
Eagle eyes, you have.  Thanks.

cindyli commented on gpii/node_modules/flowManager/test/SolutionsRegistryDataSourceTests.js:
```diff
@@ -0,0 +1,316 @@
+            }
+        }
+    },
+    modules: [{
```
Please also verify `solutionsRegistryReady` is fired. This comment applies to all test cases.

klown replied:
Okay.  I did this by adding a listener to the test SRDS component in the `testCaseHolder` because the event happens long before the test sequence starts up, and listening at that point always misses it.  I hope that's reasonable.

cindyli commented on gpii/node_modules/flowManager/test/configs/gpii.flowManager.tests.browserChannel.config.json5:
```diff
@@ -18,6 +18,12 @@
                 },
                 "target": "{that deviceReporter platformReporter}.options.invokers.reportPlatform"
+            },
+            "browserChannel.tests.repositorySolutionsLoader": {
```
Is this distribution still needed as now the SR remote fetch is controlled by an add-on config? The default behavior for all tests should be the same as having  SR remote fetch disabled.

klown replied:
Right -- removed.  But, because these test definitions are shared with the acceptance tests in `tests/UntrustedBrowserChannelTests.js`, its config file needed a similar modifiction.

cindyli commented on gpii/node_modules/flowManager/test/configs/gpii.flowManager.tests.capture.fakeData.config.json5:
```diff
@@ -11,6 +12,17 @@
                 "target": "{that flowManager solutionsRegistryDataSource}.options.path",
                 "priority": "after:flowManager.solutions"
             },
+            "capture.loadRepositorySolutions": {
```
The same comment above applies here too.

klown replied:
The distribution of a mock load-from-repository function isn't needed here in any absolute sense; however, in this case, the load registries sequence is in fact done properly, but as a mockup and it doesn't fail nor log any extra error messages.  The capture tests thereby show that the local SRDS works as described.  I thought this was relevant since the capture tool would be using the local SRDS in reality.  However, if you think this is overkill, I'll remove it.

cindyli commented on gpii/node_modules/flowManager/test/shared/BrowserChannelTestDefs.js:
```diff
@@ -53,6 +53,12 @@ gpii.tests.flowManager.browserChannel.reportPlatform = function () {
 
+// A no-op to gracefully fail fetching the solutions registry from the source
+// code repository in order to use the canned registry set up for these tests.
+gpii.tests.flowManager.browserChannel.getRepositorySolutions = function (/*repositorySolutionsLoader*/) {
```
Please double check if this function is still needed. Thanks.

klown replied:
Since it was removed from the config above, it can be removed here too.

cindyli commented:
When the add-on config that controls the SR remote fetch is ready, please write a test for it. Thanks.

klown commented:
> When the add-on config that controls the SR remote fetch is ready, please write a test for it. Thanks.

I'm not sure what you mean -- the configuration is used in the production tests where the complete local SRDS load sequence runs.  Is that not sufficient?  Should it be used elsewhere?

cindyli commented:
> > When the add-on config that controls the SR remote fetch is ready, please write a test for it. Thanks.
> 
> I'm not sure what you mean -- the configuration is used in the production tests where the complete local SRDS load sequence runs. Is that not sufficient? Should it be used elsewhere?

If the production test is adjusted to verify the SR is fetched successfully from the remote, the coverage should be sufficient.

klown commented:
> > > When the add-on config that controls the SR remote fetch is ready, please write a test for it. Thanks.
> >
> >  I'm not sure what you mean -- the configuration is used in the production tests where the complete local SRDS load sequence runs. Is that not sufficient? Should it be used elsewhere?
>
> If the production test is adjusted to verify the SR is fetched successfully from the remote, the coverage should be sufficient.

Oh, I see.  Yes, I'll put a test in there somewhere.

There is one wrinkle though (isn't there always?  :-P) ).  The `vagrantCloudBasedContainers.sh` script generates a revision by using the `HEAD` of the current branch.  This is fine for CI, but not necessarily for local VMs.  The problem is that during  development, that revision doesn't exist on github unless the changes is pushed.  If a developer runs `vagrantCloudBasedContainers.sh` before pushing their changes, which is common, then the local SRDS will make the request of github for a non-existent revision, and will fall back and use the file-based registries only.

I'll look into other git magic to get a reasonable revision.

klown commented:
@cindyli  Regarding a test of the new add-on configuration, I've added a `SolutionsRegistryLoadSequenceTests.js` to the production tests.  It uses the test environment that you developed for waiting until the system is ready before running the tests.  What is tested in the new tests is:
* that the correct revision was used by the local SRDS by testing the revision used against the one stored at the root of the docker container,
* that the local file system solutions registries were loaded (tests for non-null),
* that the remote solutions registry was loaded (tests that it's non-null)

Is that sufficient, do you think?

There are still two issues:  the test stores a reference to the local SRDS in an override of the `getRevision` step. I tried to use a `distributeOptions` to get that reference, but I couldn't figure that out.  Also, I'm still working on the issue [the issue I described above](https://github.com/GPII/universal/pull/840#issuecomment-615258538) about guaranteeing an actual github revision sha256 in all cases.

cindyli commented on package.json:
```diff
@@ -76,7 +76,7 @@
         "test:vagrantNode": "vagrant ssh -c 'cd /home/vagrant/sync/universal; npm run test:node'",
         "test:vagrantProduction": "vagrant ssh -c 'cd /home/vagrant/sync/universal; DISPLAY=:0; ./scripts/vagrantCloudBasedContainers.sh'",
-        "test:productionConfig": "node tests/production/AddUserSettingsToCouchTests.js && node tests/production/all-tests.js && node tests/production/DeleteUserSettingsFromCouchTests.js",
+        "test:productionConfig": "node tests/production/AddUserSettingsToCouchTests.js && node tests/production/all-tests.js && node tests/production/DeleteUserSettingsFromCouchTests.js && node tests/production/SolutionsRegistryLoadSequenceTests.js",
```
Can you add `tests/production/SolutionsRegistryLoadSequenceTests.js` to `tests/production/all-tests.js`?

klown replied:
Probably not.  There is a discussion on the architecture list about the security of our GCP clusters and these production tests.  The `SolutionsRegistryLoadSequenceTests` will always fail when run on GCP because GET requests outside of the cluster are not allowed.  If no special exception is made to allow the github request, it is pointless to run the loading sequence tests in GCP.  That's why I took it out of `all-tests.js`, and am adjusting the GCP yaml script to run all the other production tests except the `SolutionsRegistryLoadSequenceTests`.

klown replied:
Update:  Stepan has outlined a way to allow outbound requests to github.  I'll see if I can make that work.  If it works, then I'll put it back into `all-tests.js`

klown commented:
Tested in [GCP dev cluster](https://github.com/gpii-ops/gpii-infra/pull/610), and with @stepanstipl's pending [work-in-progress changes](https://github.com/gpii-ops/gpii-infra/pull/613), all production tests run under GCP (and pass).

Ready for review.

klown commented:
ok to test

klown commented:
ok to test
